### PR TITLE
add "final void Function()? onCloseGalleryPressed;" and  "final bool …

### PR DESCRIPTION
**Pull request by leeyisoft**
add "final void Function()? onCloseGalleryPressed;" and  "final bool showGalleryCloseButton;" for ./lib/src/widgets/chat.dart

### What does it do?

Describe the technical changes you did.

### Why is it needed?

Describe the issue you are solving.

### How to test it?

Provide information about the environment and the path to verify the behavior.

### Related issues/PRs

Let us know if this is related to any issue/pull request.

### CODE REVIEW

File: lib/src/widgets/text_message.dart [hunk 1]
```diff
@@ -152,8 +155,9 @@ class TextMessage extends StatelessWidget {
               ),
               MatchText(
                 onTap: (url) async {
-                  if (await canLaunch(url)) {
-                    await launch(url);
+                  final uri = Uri.dataFromString(url);
```
Comment (felixgabler) [inline lib/src/widgets/text_message.dart:158]: Did you test this? This method creates a `data` URI from some content. The correct way to create a launchable URI would be `Uri.parse` or, even better, `Uri.tryParse`.

Comment (felixgabler) [reply lib/src/widgets/text_message.dart:158]: I fixed a different issue in #261 that also performs this change

**Comment by felixgabler** [comment]
Perhaps we can also fix #259 in this PR?

File: lib/src/widgets/text_message.dart (hunk 1 above)
Comment (leeyisoft) [reply lib/src/widgets/text_message.dart:158]: > Did you test this? This method creates a `data` URI from some content. The correct way to create a launchable URI would be `Uri.parse` or, even better, `Uri.tryParse`.

I'm sorry. I didn't do a rigorous test！

Sorry again, I want to merge this commit into the main branch, how do I do that?

My understanding of this method (uri.dataFromString /1-5) is wrong, I have seen your code, is more rigorous approach

```
                onTap: (urlText) async {
                  final protocolIdentifierRegex = RegExp(
                    r'^((http|ftp|https):\/\/)',
                    caseSensitive: false,
                  );
                  if (!urlText.startsWith(protocolIdentifierRegex)) {
                    urlText = "https://" + urlText;
                  }
                  final url = Uri.tryParse(urlText);
                  if (url != null && await canLaunchUrl(url)) {
                    await launchUrl(url, mode: LaunchMode.externalApplication);
                  }
                },
```

**Comment by leeyisoft** [comment]
> Perhaps we can also fix #259 in this PR?

I'll copy your code and make a commit, okay?（I am pushing after receiving your reply）

```
                onTap: (urlText) async {
                  final protocolIdentifierRegex = RegExp(
                    r'^((http|ftp|https):\/\/)',
                    caseSensitive: false,
                  );
                  if (!urlText.startsWith(protocolIdentifierRegex)) {
                    urlText = "https://" + urlText;
                  }
                  final url = Uri.tryParse(urlText);
                  if (url != null && await canLaunchUrl(url)) {
                    await launchUrl(url, mode: LaunchMode.externalApplication);
                  }
                },
```

File: lib/src/widgets/chat.dart [hunk 2]
```diff
@@ -409,7 +419,15 @@ class _ChatState extends State<Chat> {
         messageWidth: _messageWidth,
         nameBuilder: widget.nameBuilder,
         onAvatarTap: widget.onAvatarTap,
-        onMessageDoubleTap: widget.onMessageDoubleTap,
+        // onMessageDoubleTap: widget.onMessageDoubleTap,
+        onMessageDoubleTap: (context, tappedMessage) {
```
[the remark this answers is not in the document]
Comment (leeyisoft) [reply lib/src/widgets/chat.dart:423]: > What is this exactly?

It was caused by my careless operation. I will close this PR immediately

[the remark this answers is not in the document]
Comment (demchenkoalex) [reply lib/src/widgets/chat.dart:423]: Hmm, but the callback/show close button changes were fine. You sure you don't want them?

[the remark this answers is not in the document]
Comment (leeyisoft) [reply lib/src/widgets/chat.dart:423]: When I have time, I will do a PR after rearranging

**Pull request closed without merging**